Foundation/URLSession/ftp/FTPURLProtocol.swift

```diff
@@ -0,0 +1,148 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+
+import CoreFoundation
+import Dispatch
+
+internal class _FTPURLProtocol: _NativeProtocol {
+    
+    public required init(task: URLSessionTask, cachedResponse: CachedURLResponse?, client: URLProtocolClient?) {
+        super.init(task: task, cachedResponse: cachedResponse, client: client)
+    }
+    
+    public required init(request: URLRequest, cachedResponse: CachedURLResponse?, client: URLProtocolClient?) {
+        super.init(request: request, cachedResponse: cachedResponse, client: client)
+    }
+    
+    override class func canInit(with request: URLRequest) -> Bool {
+        guard request.url?.scheme == "ftp" || request.url?.scheme == "ftps" || request.url?.scheme == "sftp" else { return false }
+        return true
+    }
+    
+    override class func canonicalRequest(for request: URLRequest) -> URLRequest {
+        return request
+    }
+    
+    override func startLoading() {
+        resume()
+    }
+    
+    override func stopLoading() {
+        if task?.state == .suspended {
+            suspend()
+        } else {
+            self.internalState = .transferFailed
+            guard let error = self.task?.error else { fatalError() }
+            completeTask(withError: error)
+        }
+    }
+    
+    override  func didReceive(headerData data: Data, contentLength: Int64) -> _EasyHandle._Action {
+        guard case .transferInProgress(let ts) = internalState else { fatalError("Received body data, but no transfer in progress.") }
+        guard let task = task else { fatalError("Received header data but no task available.") }
+        task.countOfBytesExpectedToReceive = contentLength > 0 ? contentLength : NSURLSessionTransferSizeUnknown
+        do {
+            let newTS = try ts.byAppendingFTP(headerLine: data, expectedContentLength: contentLength)
+            internalState = .transferInProgress(newTS)
+            let didCompleteHeader = !ts.isHeaderComplete && newTS.isHeaderComplete
+            if didCompleteHeader {
+                // The header is now complete, but wasn't before.
+                didReceiveResponse()
+            }
+            return .proceed
+        } catch {
+            return .abort
+        }
+    }
+    
+    override func seekInputStream(to position: UInt64) throws {
+        // NSUnimplemented()
+    }
+    
+    override  func updateProgressMeter(with propgress: _EasyHandle._Progress) {
+        //NSUnimplemented()
+    }
+    
+    override func configureEasyHandle(for request: URLRequest) {
+        easyHandle.set(verboseModeOn: enableLibcurlDebugOutput)
+        easyHandle.set(debugOutputOn: enableLibcurlDebugOutput, task: task!)
+        easyHandle.set(skipAllSignalHandling: true)
+        guard let url = request.url else { fatalError("No URL in request.") }
```

Comment:
As previous - check files for `else { fatalError` and indent lines as appropriate